docs/src/explanations/properties.md

```diff
@@ -148,3 +148,249 @@ on integral `Property` typed values.
+with the `Definition` and `Instance` APIs.
+
+The readable fields we want to expose on `Object`s of the `CSRDescription` class
+are a string identifier, a string description, and an integer bitwidth, so these
+are output `Property` type ports on the `Class`.
+
+To capture concrete values at each `Object` instantiation, we have corresponding
+input `Property` type ports, which are connected directly to the outputs. This
+is how we would represent something like a Scala `case class` using `Class`.
+
+```scala mdoc:silent
+// A hardware module representing a CSR and its description.
+class CSRModule(
+  csrDescDef:     Definition[CSRDescription],
+  width:          Int,
+  identifierStr:  String,
+  descriptionStr: String)
+    extends Module {
+  override def desiredName = identifierStr
+
+  // Create a hardware port for the CSR value.
+  val value = IO(Output(UInt(width.W)))
+
+  // Create a property port for a reference to the CSR description object.
+  val description = IO(Output(csrDescDef.getPropertyType))
+
+  // Instantiate a CSR description object, and connect its input properties.
+  val csrDescription = Instance(csrDescDef)
+  csrDescription.identifierIn := Property(identifierStr)
+  csrDescription.descriptionIn := Property(descriptionStr)
+  csrDescription.widthIn := Property(width)
+
+  // Create a register for the hardware CSR. A real implementation would be more involved.
+  val csr = RegInit(0.U(width.W))
+
+  // Assign the CSR value to the hardware port.
+  value := csr
+
+  // Assign a reference to the CSR description object to the property port.
+  description := csrDescription.getPropertyReference
+}
+```
+
+The `CSRModule` is a `Module` that represents the (dummy) hardware for a CSR, as
+well as a `CSRDescription`. Using a `Definition` of a `CSRDescription`, an
+`Object` is created and its inputs supplied from the `CSRModule` constructor
+arguments. Then, a reference to the `Object` is connected to the `CSRModule`
+output, so the reference will be exposed to the outside.
+
+```scala mdoc:silent
+// The entrypoint module.
+class Top extends Module {
+  // Create a Definition for the CSRDescription Class.
+  val csrDescDef = Definition(new CSRDescription)
+
+  // Get the CSRDescription ClassType.
+  val csrDescType = csrDescDef.getClassType
+
+  // Create a property port to collect all the CSRDescription object references.
+  val descriptions = IO(Output(Property[Seq[csrDescType.Type]]()))
+
+  // Instantiate a couple CSR modules.
+  val mcycle = Module(new CSRModule(csrDescDef, 64, "mcycle", "Machine cycle counter."))
+  val minstret = Module(new CSRModule(csrDescDef, 64, "minstret", "Machine instructions-retired counter."))
+
+  // Assign references to the CSR description objects to the property port.
+  descriptions := Property(Seq(mcycle.description.as(csrDescType), minstret.description.as(csrDescType)))
+}
+```
+
+The `Top` module represents the entrypoint. It creates the `Definition` of the
```

Comment:
The `Definition` usage makes sense to me, but w/o the D/I API, does it also works?

Reply:
Yep. I haven't show an example using the "unsafe" APIs, but it's possible to create classes and objects without D/I. I wanted the first example to show the "safe" APIs with D/I, but I can add a follow up showing the unsafe version.

Reply:
I'll add some description about why we recommend Definition/Instance, and just mention DynamicObject, which has scaladoc for the unsafe APIs. We can add an example using the unsafe APIs separately if it's helpful.

Reply:
Nice!